src/main/java/org/fogbowcloud/manager/core/plugins/compute/openstack/OpenStackNovaV2ComputePlugin.java

```diff
@@ -0,0 +1,438 @@
+    private static final String KEY_JSON_FIELD = "key_name";
+    private static final String PUBLIC_KEY_JSON_FIELD = "public_key";
+    private static final String KEYPAIR_JSON_FIELD = "keypair";
+    private static final String UUID_JSON_FIELD = "uuid";
+    private static final String FOGBOW_INSTANCE_NAME = "fogbow-instance-";
+    private static final String TENANT_ID = "tenantId";
+    private static final String SERVERS = "/servers";
+    private static final String SUFFIX_ENDPOINT_KEYPAIRS = "/os-keypairs";
+    private static final String SUFFIX_ENDPOINT_FLAVORS = "/flavors";
+    private static final String COMPUTE_NOVAV2_URL_KEY = "compute_novav2_url";
+    private static final String NO_VALID_HOST_WAS_FOUND = "No valid host was found";
+    private final String COMPUTE_V2_API_ENDPOINT = "/v2/";
+
+    private static final String COMPUTE_NOVAV2_NETWORK_KEY = "compute_novav2_network_id";
+
+    private static final Logger LOGGER = Logger.getLogger(OpenStackNovaV2ComputePlugin.class);
+
+    private List<Flavor> flavors;
+    private Properties properties;
+    private HttpClient client;
+
+    public OpenStackNovaV2ComputePlugin(Properties properties) {
+        this.flavors = new ArrayList<>();
+        this.properties = properties;
+        this.initClient();
+    }
+
+    public String requestInstance(ComputeOrder computeOrder, String imageId) throws RequestException {
+        LOGGER.debug("Requesting instance with token=" + computeOrder.getLocalToken());
+
+        Token localToken = computeOrder.getLocalToken();
+        String flavorId = updateAndFindSmallestFlavor(computeOrder).getId();
+        String tenantId = getTenantId(localToken);
+        String networkId = getNetworkId();
+        String userData = computeOrder.getUserData().getContent();
+        String keyName = getKeyName(tenantId, localToken, computeOrder.getPublicKey());
+        String endpoint = getComputeEndpoint(tenantId, SERVERS);
+
+        try {
+            JSONObject json = generateJsonRequest(imageId, flavorId, userData, keyName, networkId);
+            String jsonResponse = doPostRequest(endpoint, localToken, json);
+
+            return getAttFromJson(ID_JSON_FIELD, jsonResponse);
+        } catch (JSONException e) {
+            LOGGER.error(e);
+            throw new RequestException(ErrorType.BAD_REQUEST, ResponseConstants.IRREGULAR_SYNTAX);
+        } finally {
+            if (keyName != null) {
+                try {
+                    deleteKeyName(tenantId, localToken, keyName);
+                } catch (Throwable t) {
+                    LOGGER.warn("Could not delete key.", t);
+                }
+            }
+        }
+    }
+
+    private void initClient() {
+        HttpRequestUtil.init(properties);
+        this.client = HttpRequestUtil.createHttpClient();
+    }
+
+    protected String getTenantId(Token localToken) {
+        return localToken.getAttributes().get(TENANT_ID);
```

Comment:
These methods chaining calls are bad for debug.

Reply:
Fixed!